src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/CodeStyle/CodeStyleOption2`1.cs

```diff
@@ -0,0 +1,214 @@
+
+            return new CodeStyleOption2<T>(value, severity switch
+            {
+                DiagnosticSeverity.Hidden => NotificationOption2.Silent,
+                DiagnosticSeverity.Info => NotificationOption2.Suggestion,
+                DiagnosticSeverity.Warning => NotificationOption2.Warning,
+                DiagnosticSeverity.Error => NotificationOption2.Error,
+                _ => throw new ArgumentException(nameof(element)),
+            });
+        }
+
+        private static Func<string, T> GetParser(string type)
+            => type switch
+            {
+                nameof(Boolean) =>
+                    // Try to map a boolean value.  Either map it to true/false if we're a 
+                    // CodeStyleOption<bool> or map it to the 0 or 1 value for an enum if we're
+                    // a CodeStyleOption<SomeEnumType>.
+                    (Func<string, T>)(v => Convert(bool.Parse(v))),
+                nameof(Int32) => v => Convert(int.Parse(v)),
+                nameof(String) => v => (T)(object)v,
+                _ => throw new ArgumentException(nameof(type)),
+            };
+
+        private static T Convert(bool b)
+        {
+            // If we had a bool and we wanted a bool, then just return this value.
+            if (typeof(T) == typeof(bool))
+            {
+                return (T)(object)b;
+            }
+
+            // Map booleans to the 1/0 value of the enum.
+            return b ? (T)(object)1 : (T)(object)0;
+        }
+
+        private static T Convert(int i)
+        {
+            // We got an int, but we wanted a bool.  Map 0 to false, 1 to true, and anything else to default.
+            if (typeof(T) == typeof(bool))
+            {
+                return (T)(object)(i == 1);
+            }
+
+            // If had an int and we wanted an enum, then just return this value.
+            return (T)(object)(i);
+        }
+
+        public bool Equals(CodeStyleOption2<T> other)
+            => EqualityComparer<T>.Default.Equals(Value, other.Value) &&
+               Notification == other.Notification;
+
+        public override bool Equals(object obj)
+            => obj is CodeStyleOption2<T> option &&
+               Equals(option);
+
+        public override int GetHashCode()
+            => unchecked((Notification.GetHashCode() * (int)0xA5555529) + Value.GetHashCode());
+
+#if !CODE_STYLE
+        public static implicit operator CodeStyleOption<T>(CodeStyleOption2<T> option)
+        {
+            if (option == null)
+            {
+                return null;
+            }
+
+            return new CodeStyleOption<T>(option.Value, option.Notification);
+        }
+
+        public static implicit operator CodeStyleOption2<T>(CodeStyleOption<T> option)
+        {
+            if (option == null)
+            {
+                return null;
+            }
+
+            return new CodeStyleOption2<T>(option.Value, option.Notification);
```

Comment:
❕  If `CodeStyleOption<T>` is a wrapper for `CodeStyleOption2<T>`, this operator should unwrap the underlying value instead of trying to construct a new one.

Reply:
Good point - let me change.

Reply:
Now this is no longer true - both of them are based off an OptionDefinition, which is indeed passed through.